library.rb

```diff
@@ -1,3 +1,129 @@
+require './author'
+require './book'
+require './reader'
+require './order'
+
 class Library
+  REQUIRED_COLLECTIONS = %w(Author Book Reader Order)
+  attr_reader :authors, :books, :readers, :orders
+  def initialize
+    @authors = []
+    @books = []
+    @readers = []
+    @orders = []
+    puts 'Use :load method for load library from file.'
+  end
+
+  def load
+    REQUIRED_COLLECTIONS.each do |name|
+      load_from_file(name)
+    end
+  end
+
+  def save
+    REQUIRED_COLLECTIONS.each do |name|
+      save_to_file(name)
+    end
+  end
+
+# for interface
+  def new_author(params)
+    new_class_instance("Author", @authors, params)
+  end
+
+  def new_book(params)
+    unless @authors.map(&:name).index(params.last)
+      puts "Library does not have the author #{params.last}"
+      return
+    end
+    new_class_instance("Book", @books, params)
+  end
+
+  def new_reader(params)
+    new_class_instance("Reader", @readers, params)
+  end
+
+  def new_order(params)
+    unless @books.map(&:title).index(params.first)
+      puts "Library does not have the book #{params.first}"
+      return
+    end
+    unless @readers.map(&:name).index(params[1])
+      puts "#{params[1]} wasn't registered at this library"
+      return
+    end
+    new_class_instance("Order", params)
+  end  
+
+  def best_reader
+    readers = @orders.map(&:reader)
+    uniq_readers = readers.uniq
+    uniq_readers.sort! { |x,y| readers.count(y) <=> readers.count(x) }
+    # in case when we have few readers with equel quantity read books
+    # example in the order txt file
+    # Best reader is(are):
+    # Andrey Kraveckiy
+    # Sergey Evtushenko
+    uniq_readers.select { |e| readers.count(e) == readers.count(uniq_readers.first) }
+  end
+
+  def popular_book
+    books = @orders.map(&:book)
+    uniq_books = books.uniq
+    uniq_books.sort! { |x,y| books.count(y) <=> books.count(x) }
+    # in case when we have few books with equel quantity in orders
+    # example in the order txt file
+    # Popular book is(are):
+    # Ruslan and Liudmila
+    # People like a goods
+    # About person, life and nature
+
+    uniq_books.select { |e| books.count(e) == books.count(uniq_books.first) }
+  end
+
+  def the_3_most_popular
+    books = @orders.map(&:book)
+    uniq_books = books.uniq
+    uniq_books.sort! { |x,y| books.count(y) <=> books.count(x) }
+    answer = []
+    uniq_books.first(3).each do |book|
+      answer += @orders.map { |ord| ord.reader if ord.book == book }
+    end
+    answer.compact.uniq.length
```

Comment:
1. You can use modified method popular_book with parameter 3 to get popular books. Then select orders with these books, group them by reader and count 
2. You can group orders by book, sort by their orders count, get orders, group by reader, count